File: packages/lsp/src/manifest.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import { MANIFEST_FILENAME } from "./constants.js";
import type { LspManifest, LspServerProfile } from "./types.js";

/**
 * Load an .lsp.json manifest from a workspace root, or from an explicit
 * path. Returns undefined if the file doesn't exist — that's not an
 * error; the orchestrator then returns SERVER_NOT_AVAILABLE on the
 * first operation that needs a server.
 *
 * Validation is permissive: we accept the Claude Code manifest shape
 * and reject entries that look malformed (e.g. missing command).
 * Parse errors throw.
 */
export async function loadManifest(
  explicitPath: string | undefined,
  workspaceRoot: string,
): Promise<LspManifest | undefined> {
  const manifestPath = explicitPath ?? path.join(workspaceRoot, MANIFEST_FILENAME);
  let text: string;
  try {
    text = await readFile(manifestPath, "utf8");
  } catch {
    return undefined;
  }
  const parsed = JSON.parse(text) as unknown;
  return normalizeManifest(parsed, manifestPath);
}

function normalizeManifest(
  raw: unknown,
  source: string,
): LspManifest {
  if (raw === null || typeof raw !== "object") {
    throw new Error(`Invalid LSP manifest at ${source}: expected object`);
  }
  const servers = (raw as { servers?: unknown }).servers;
  if (servers === undefined || servers === null || typeof servers !== "object") {
    throw new Error(
      `Invalid LSP manifest at ${source}: expected { servers: { ... } }`,
    );
  }
  const out: Record<string, LspServerProfile> = {};
  for (const [key, value] of Object.entries(
    servers as Record<string, unknown>,
  )) {
    const profile = normalizeProfile(key, value, source);
    out[key] = profile;
  }
  return { servers: out };
}

function normalizeProfile(
  name: string,
  raw: unknown,
  source: string,
): LspServerProfile {
  if (raw === null || typeof raw !== "object") {
    throw new Error(
      `Invalid LSP server profile '${name}' in ${source}: expected object`,
    );
  }
  const r = raw as {
    extensions?: unknown;
    command?: unknown;
    rootPatterns?: unknown;
    initializationOptions?: unknown;
  };
  const extensions = Array.isArray(r.extensions)
    ? (r.extensions as string[]).filter((x) => typeof x === "string")
    : [];
  if (extensions.length === 0) {
    throw new Error(
      `LSP server '${name}' in ${source}: 'extensions' must be a non-empty string[]`,
    );
  }
  const command = Array.isArray(r.command)
    ? (r.command as string[]).filter((x) => typeof x === "string")
    : [];
  if (command.length === 0) {
    throw new Error(
      `LSP server '${name}' in ${source}: 'command' must be a non-empty string[]`,
    );
  }
  const rootPatterns = Array.isArray(r.rootPatterns)
    ? (r.rootPatterns as string[]).filter((x) => typeof x === "string")
    : undefined;
  const initializationOptions =
    r.initializationOptions !== undefined &&
    typeof r.initializationOptions === "object" &&
    r.initializationOptions !== null
      ? (r.initializationOptions as Record<string, unknown>)
      : undefined;
  return {
    language: name,
    extensions,
    command,
    ...(rootPatterns !== undefined ? { rootPatterns } : {}),
    ...(initializationOptions !== undefined ? { initializationOptions } : {}),
  };
}

/**
 * Given a filesystem path and a manifest, return the matching server
 * profile based on extension. Returns undefined if no match.
 */
export function profileForPath(
  filePath: string,
  manifest: LspManifest | undefined,
): LspServerProfile | undefined {
  if (manifest === undefined) return undefined;
  const ext = path.extname(filePath).toLowerCase();
  if (ext.length === 0) return undefined;
  for (const profile of Object.values(manifest.servers)) {
    if (profile.extensions.some((e) => e.toLowerCase() === ext)) {
      return profile;
    }
  }
  return undefined;
}

/**
 * Find the LSP root for a server profile by walking up from the file
 * until we find one of `rootPatterns`. Falls back to the workspace cwd
 * if no pattern matches.
 */
export async function findLspRoot(
  filePath: string,
  profile: LspServerProfile,
  workspaceCwd: string,
): Promise<string> {
  if (!profile.rootPatterns || profile.rootPatterns.length === 0) {
    return workspaceCwd;
  }
  const { access } = await import("node:fs/promises");
  let current = path.dirname(filePath);
  while (current.length > 1) {
    for (const pattern of profile.rootPatterns) {
      const candidate = path.join(current, pattern);
      try {
        await access(candidate);
        return current;
      } catch {
        // not this dir; keep walking
      }
    }
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return workspaceCwd;
}

File: crates/lsp/src/manifest.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::constants::MANIFEST_FILENAME;
use crate::types::{LspManifest, LspServerProfile};

/// Load an .lsp.json manifest from `explicit_path` if given, else from
/// `<workspace_root>/.lsp.json`. Returns Ok(None) when the file is
/// missing (that's not an error; orchestrator returns SERVER_NOT_AVAILABLE).
/// Parse / shape errors are Err(msg).
pub async fn load_manifest(
    explicit_path: Option<&str>,
    workspace_root: &str,
) -> Result<Option<LspManifest>, String> {
    let path: PathBuf = match explicit_path {
        Some(p) => PathBuf::from(p),
        None => Path::new(workspace_root).join(MANIFEST_FILENAME),
    };
    let text = match tokio::fs::read_to_string(&path).await {
        Ok(t) => t,
        Err(_) => return Ok(None),
    };
    let raw: serde_json::Value = serde_json::from_str(&text)
        .map_err(|e| format!("Invalid JSON in {}: {}", path.display(), e))?;
    let manifest = normalize_manifest(&raw, &path.display().to_string())?;
    Ok(Some(manifest))
}

fn normalize_manifest(raw: &serde_json::Value, source: &str) -> Result<LspManifest, String> {
    let obj = raw
        .as_object()
        .ok_or_else(|| format!("Invalid LSP manifest at {}: expected object", source))?;
    let servers = obj
        .get("servers")
        .and_then(|v| v.as_object())
        .ok_or_else(|| {
            format!(
                "Invalid LSP manifest at {}: expected {{ servers: {{ ... }} }}",
                source
            )
        })?;
    let mut out: HashMap<String, LspServerProfile> = HashMap::new();
    for (key, value) in servers {
        let profile = normalize_profile(key, value, source)?;
        out.insert(key.clone(), profile);
    }
    Ok(LspManifest { servers: out })
}

fn normalize_profile(
    name: &str,
    raw: &serde_json::Value,
    source: &str,
) -> Result<LspServerProfile, String> {
    let obj = raw.as_object().ok_or_else(|| {
        format!(
            "Invalid LSP server profile '{}' in {}: expected object",
            name, source
        )
    })?;
    let extensions: Vec<String> = obj
        .get("extensions")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default();
    if extensions.is_empty() {
        return Err(format!(
            "LSP server '{}' in {}: 'extensions' must be a non-empty string[]",
            name, source
        ));
    }
    let command: Vec<String> = obj
        .get("command")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default();
    if command.is_empty() {
        return Err(format!(
            "LSP server '{}' in {}: 'command' must be a non-empty string[]",
            name, source
        ));
    }
    let root_patterns = obj.get("rootPatterns").and_then(|v| v.as_array()).map(|arr| {
        arr.iter()
            .filter_map(|x| x.as_str().map(|s| s.to_string()))
            .collect()
    });
    let initialization_options = obj.get("initializationOptions").cloned();
    Ok(LspServerProfile {
        language: name.to_string(),
        extensions,
        command,
        root_patterns,
        initialization_options,
    })
}

pub fn profile_for_path(
    file_path: &str,
    manifest: Option<&LspManifest>,
) -> Option<LspServerProfile> {
    let Some(manifest) = manifest else {
        return None;
    };
    let ext = Path::new(file_path)
        .extension()
        .and_then(|s| s.to_str())
        .map(|s| format!(".{}", s.to_ascii_lowercase()))?;
    for profile in manifest.servers.values() {
        if profile
            .extensions
            .iter()
            .any(|e| e.to_ascii_lowercase() == ext)
        {
            return Some(profile.clone());
        }
    }
    None
}

pub async fn find_lsp_root(
    file_path: &str,
    profile: &LspServerProfile,
    workspace_cwd: &str,
) -> String {
    let Some(patterns) = &profile.root_patterns else {
        return workspace_cwd.to_string();
    };
    if patterns.is_empty() {
        return workspace_cwd.to_string();
    }
    let mut current = Path::new(file_path)
        .parent()
        .map(|p| p.to_path_buf())
        .unwrap_or_else(|| PathBuf::from("/"));
    loop {
        for pattern in patterns {
            let candidate = current.join(pattern);
            if tokio::fs::metadata(&candidate).await.is_ok() {
                return current.to_string_lossy().into_owned();
            }
        }
        let parent = current.parent().map(|p| p.to_path_buf());
        match parent {
            Some(p) if p != current => current = p,
            _ => break,
        }
    }
    workspace_cwd.to_string()
}
